tests/misc/test_dmc.py

```diff
@@ -0,0 +1,124 @@
+        # memory-actual may not equal memory-target even if ballooning has finished.
+        # Use a heuristic instead, where ballooning is determined to have finished if memory-actual is within 1% of
+        # memory-target.
+        return abs(self._memory_actual - self._memory_target) < self._memory_target // 100
+
+    def poll(self):
+        """Determine if ballooning has finished or stalled."""
+        self._update()
+        return self.is_memory_target_satisfied() or time.perf_counter() - self._timestamp > 10.0
+
+    def memory_actual(self):
+        return self._memory_actual
+
+
+def wait_for_vm_balloon_finished(vm: VM):
+    memory_target = int(vm.param_get("memory-target"))
+    logging.info("Wait for ballooning to finish")
+    tracker = DmcMemoryTracker(vm, memory_target)
+    wait_for(tracker.poll, timeout_secs=300)
+    logging.info(f"Current memory: {tracker.memory_actual()}")
+    assert tracker.is_memory_target_satisfied()
+
+
+@pytest.fixture(scope="module")
+def imported_vm_and_snapshot(imported_vm: VM):
+    vm = imported_vm
+    snapshot = vm.snapshot()
+    yield vm, snapshot
+    snapshot.destroy(verify=True)
+
+
+@pytest.fixture
+def vm_with_memory_limits(imported_vm_and_snapshot: Tuple[VM, Snapshot]):
+    vm, snapshot = imported_vm_and_snapshot
+    vm.set_memory_limits(static_max=MEMORY_TARGET_HIGH, dynamic_min=MEMORY_TARGET_HIGH, dynamic_max=MEMORY_TARGET_HIGH)
+    snapshot = vm.snapshot()
+    yield vm
+    snapshot.revert()
+
+
+@pytest.mark.small_vm
+class TestDmc:
+    def start_dmc_vm(self, vm: VM):
+        vm.start()
+        vm.wait_for_vm_running_and_ssh_up()
+        if vm.param_get("other", "feature-balloon", accept_unknown_key=True) != "1":
+            pytest.skip("VM does not support ballooning")
+
+    def test_dmc_start_low(self, vm_with_memory_limits: VM):
+        """Start the VM with less memory than the static max."""
+        vm = vm_with_memory_limits
+        vm.set_memory_target(MEMORY_TARGET_LOW)
+        self.start_dmc_vm(vm)
+        wait_for_vm_balloon_finished(vm)
+        # restore
+        vm.set_memory_target(MEMORY_TARGET_HIGH)
+        wait_for_vm_balloon_finished(vm)
+
+    def test_dmc_decrease(self, vm_with_memory_limits: VM):
+        """Decrease the memory of a VM that started without DMC."""
+        vm = vm_with_memory_limits
+        self.start_dmc_vm(vm)
+        vm.set_memory_target(MEMORY_TARGET_LOW)
+        wait_for_vm_balloon_finished(vm)
+        # restore
+        vm.set_memory_target(MEMORY_TARGET_HIGH)
+        wait_for_vm_balloon_finished(vm)
+
+    def test_dmc_suspend_pod(self, vm_with_memory_limits: VM):
+        """Suspend a VM with DMC and populate-on-demand enabled."""
+        vm = vm_with_memory_limits
+        vm.set_memory_target(MEMORY_TARGET_LOW)
+        self.start_dmc_vm(vm)
+        wait_for_vm_balloon_finished(vm)
+        vm.suspend(verify=True)
+        vm.resume()
+        vm.wait_for_vm_running_and_ssh_up()
```

Comment:
this test is supposed to fail sometimes, right? at least that'd be the expectation given what we know about VMs running out of PoD memory on domain_save...

Reply:
If so, perhaps worth adding comments about issues we've seen, e.g.:
```
[2025-06-23 15:29:49] (XEN) [2312934.000562] p2m_pod_demand_populate: Dom9 out of PoD memory! (tot=1048605 ents=992 dom0)
[2025-06-23 15:29:49] (XEN) [2312934.000566] domain_crash called from p2m_pod_demand_populate+0x4d2/0x8b0
```